core/arch/arm/kernel/thread_spmc.c

```diff
@@ -348,12 +300,17 @@ static bool is_nil_uuid(uint32_t w0, uint32_t w1, uint32_t w2, uint32_t w3)
 	return !w0 && !w1 && !w2 && !w3;
 }
 
-static bool is_optee_os_uuid(uint32_t w0, uint32_t w1, uint32_t w2, uint32_t w3)
+static bool is_my_uuid(uint32_t w0, uint32_t w1, uint32_t w2, uint32_t w3)
 {
-	return w0 == OPTEE_MSG_OS_OPTEE_UUID_0 &&
-	       w1 == OPTEE_MSG_OS_OPTEE_UUID_1 &&
-	       w2 == OPTEE_MSG_OS_OPTEE_UUID_2 &&
-	       w3 == OPTEE_MSG_OS_OPTEE_UUID_3;
+	/*
+	 * This depends on which UUID we have been assigned.
+	 * TODO add a generic mechanism to obtain our UUID.
```

Comment:
Once a UUID is assigned to OP-TEE it won't change I suppose, so why would a "generic" mechanism be needed?
Morover, why use a different UUID from the one that already exists in core/include/optee_msg.h? Why rename this function?

Reply:
What if there's two instances of OP-TEE? Both cannot have the same UUID.
Another problem with `OPTEE_MSG_OS_OPTEE_UUID_*` is that it uses an incorrect byte order.
For FF-A we must use a correct byte order and instead of trying to reuse `OPTEE_MSG_OS_OPTEE_UUID_*` with the confusion that would lead to we're breaking free here. I'm not sure if the UUID should be hardcoded into the OP-TEE binary or if it should be supplied in some way at boot.

Reply:
Oh I see, so that's an **instance** ID. Hmmmm. Shouldn't all this be handled by SPMD? The OP-TEE OS has no use of whatever ID has been assigned to it.

Reply:
I think we should call it **instance UUID** to avoid confusion with **partition ID**.

This is when OP-TEE is the SPMC. It should in theory be possible to use the same SPMD regardless if the SPMC is at S-EL2 or S-EL1.

I realize that this can be a bit confusing. The point is that we try minimize the impact of SPMC at different levels on EL3 and Normal World. This makes the migration path via SPMC at S-EL1 more useful as the final change can be done in Secure world only leaving other parts more or less unchanged.

Reply:
You've lost me :confused: 

Is it possible to have OP-TEE be the SPMC and still have several instances of OP-TEE (with S-EL2 underneath)?

Reply:
In all cases we have a SPMD (SPM Dispatcher) at EL3 which is more or less the same in all these configurations.

Here's the configurations we're currently aiming for:
1a. (no S-EL2) S-EL1 SPMC - OP-TEE is SPMC (and SP) at S-EL1
1b. (no S-EL2) S-EL1 SPMC - OP-TEE is SPMC (and SP) at S-EL1 with SPs at S-EL0
2a. S-EL2 SPMC - Hafnium is SPMC at S-EL2, OP-TEE is a SP at S-EL1

In 1a OP-TEE is a SPMC in the sense that it provides such an interface for SPMD and Normal World.
In 1b OP-TEE is a proper SPMC since it also manages S-EL0 SPs, this is what @jellesels-arm and company are upstreaming.
In 2a is OP-TEE just a SP, the SPMC interface towards SPMD and Normal world is provided by Hafnium

There is another one we haven't come to yet which is a bit more complicated
2b. S-EL2 SPMC - Hafnium is SPMC at S-EL2, OP-TEE is a SPMC and SP at S-EL1 with SPs at S-EL0
In 2b OP-TEE is SPMC for SPs at S-EL0 so we have a split SPMC or a two level SPMC.

To normal world all these look more or less the same. It should in fact be possible to run with an unchanged kernel at least from this point of view. This is the reason why I'd like to pick a new hard coded UUID for 1 so it can be reused in 2 without any changes to normal world.

Then there's also the byte order problem with `OPTEE_MSG_OS_OPTEE_UUID_*` which I mentioned above.

In 2b it is actually possible to have multiple OP-TEEs while at the same time having OP-TEE as SPMC.

Reply:
Thanks for the details :+1: 

> In 2b it is actually possible to have multiple OP-TEEs while at the same time having OP-TEE as SPMC.

If I understand correctly, in 2b each OP-TEE instance would have a different UUID and would need to retrieve it by some yet unspecified mechanism?

Reply:
> > In 2b it is actually possible to have multiple OP-TEEs while at the same time having OP-TEE as SPMC.
> 
> If I understand correctly, in 2b each OP-TEE instance would have a different UUID and would need to retrieve it by some yet unspecified mechanism?

Yes, and I should have mentioned that in 2a it's also possible to have multiple OP-TEE instances, it's just that they don't have any SPMC capability. However in the simplest 2a configuration (the one we're currently working on) we have only one OP-TEE instance.

Reply:
> There is another one we haven't come to yet which is a bit more complicated
> 2b. S-EL2 SPMC - Hafnium is SPMC at S-EL2, OP-TEE is a SPMC and SP at S-EL1 with SPs at S-EL0
> In 2b OP-TEE is SPMC for SPs at S-EL0 so we have a split SPMC or a two level SPMC.

Hey Jens, FWIW, there are no plans to support 2b) in the FF-A spec. A split SPMC configuration will make life more complicated for both Hafnium and OP-TEE. We are using OP-TEE as the "incubating" SPMC for S-EL0 SPs until S-EL2 systems are widely available. Once this happens, S-EL0 SPs will become peers of OP-TEE under Hafnium as the SPMC. This is not going to happen in a hurry though on production systems! If you think there is value in supporting 2b) then lets discuss.

Reply:
> Once this happens, S-EL0 SPs will become peers of OP-TEE under Hafnium as the SPMC.

What would be S-EL1 in this case?

Reply:
> What would be S-EL1 in this case?
OP-TEE in S-EL1 will be a Secure VM/Secure partition with no SPMC responsibilities. 

Since the S-EL0 SPs will use the FF-A transport and already share any protocols with OP-TEE, any communication between them will be preserved even though Hafnium will be the SPMC. 

Does this answer your question?

In fact, thinking about it a bit more, a scenario where a split SPMC might be needed is if OP-TEE TAs start using FF-A ABIs. I wonder if this is what Jens has in mind.

Reply:
>>   What would be S-EL1 in this case?
>
Got it, thanks.

Reply:
> In fact, thinking about it a bit more, a scenario where a split SPMC might be needed is if OP-TEE TAs start using FF-A ABIs. I wonder if this is what Jens has in mind.

Yes 2b would be needed for TAs using FF-A ABIs, but I don't have any plans for that in the nearest future. Once the S-EL0 stuff has landed it could be interesting to see how TAs might use the FF-A ABI instead.

I only mentioned 2b since it is a possible combination and I thought I had seen it mentioned somewhere. Thanks for the clarification.